design/decisions/0004-entity-dregister-api.md

```diff
@@ -0,0 +1,147 @@
+  For example, a device `te/device/child01//` may have an associated service `te/device/child01/service/service01`
+  and nested child devices `te/device/child11//` and `te/device/child12//`.
+  Even though the parent device and service has a common prefix `te/device/child01/#`, there is no such common prefix
+  between the parent device and its child devices.
+  And when custom topic schemes are used, even the device and its service may not have any common prefix available.
+
+## Solution Proposals
+
+### Using third-party MQTT tools like mosquitto_sub
+
+The `mosquitto_sub` command provides options like `--remove-retained` to selectively remove all retained messages in a topic hierarchy by using topic filters.
+For example, the following command can be used to remove all retained messages under the `te/device/child01` topic tree:
+
+```sh
+mosquitto_sub --remove-retained -t "te/device/child01/+/+/#"
+```
+
+The mappers must also be updated to remove entities from their entity store on receipt of these clear messages.
+
+**Pros**
+
+* Existing out-of-the-box solution.
+
+**Cons**
+
+* Works only for topic schemes that ensures that parent and child entities have a common topic prefix.
+* Even the default topic scheme ensures common topic prefix only for a device and its services and not for its nested children.
+  So, it only works for simple deployments that does not have nested child devices.
+
+### Tedge deregister command
+
+Provide a `tedge` command that takes the topic id of the entity to be deregistered as follows:
+
+```sh
+tedge deregister te/device/child01//
+```
+
+This command subscribes to all entity messages, recreates the entity store on its own so that it can identify
+the given entity and all of its linked entities so that they can all be deregistered.
+
+:::note
+Subscribing to the given entity topic tree as `te/device/child01/#` isn't enough to identify all the linked entities
+as they may not share a common topic prefix.
+:::
+
+Once all the target entities are identified, it issues secondary subscriptions for all the data topics of these entities,
+like their twin data, alarm and command topics so that all those retained message are also cleared.
+On receipt of these clear messages, the mappers must clear the corresponding entries from their own in-memory entity stores,
+and propagate that to the cloud.
+
+The following options are provided to enable more fine-grained deregistration of entities:
+* `--services-only`: Delete only the associated services of the given device and not the device itself.
+* `--children-only`: Delete only the nested child devices not the device itself.
+* `--child-type`: Delete all the child devices with the given type
+
+**Pros**
+
+* Offers better targeting of entities, than what can be done using topic filters.
+
+**Cons**
+
+* Deregistrations can't be done remotely, unless `tedge` is installed on those remote device as well.
```

Comment:
What I meant is that, if the API to deregister the entities is provided only as a `tedge` command, then it can only be invoked from a device where `tedge` is installed, even though the command internally uses MQTT only. That's why it is still possible to do this remotely as well, as long as `tedge` is installed on that remote device as well.